lib/location/cloud_service/cloud_service_here_rest.c

```diff
@@ -220,16 +310,82 @@ static int location_service_generate_request(
 		       here_adjust_rsrq(cell_data->current_cell.rsrq),
 		       here_adjust_ta(cell_data->current_cell.timing_advance),
 		       neighbors);
+
 	if ((len < 0) || (len >= buf_len)) {
-		LOG_ERR("Too small buffer for HTTP request body");
-		return -ENOMEM;
+		LOG_ERR("Too small buffer for current cell with neighbors");
+		err = -ENOMEM;
 	}
+end:
+	k_free(neighbors);
+	return err;
+}
 
-	return 0;
+
+static int cloud_service_here_rest_pos_req_encode(
+	const struct lte_lc_cells_info *cell_data,
+	const struct wifi_scan_info *wifi_data,
+	char **buf)
+{
+	int err = 0;
+	char *cell_buf = NULL;
+	char *wifi_buf = NULL;
+
+	__ASSERT_NO_MSG(cell_data != NULL || wifi_data != NULL);
+	__ASSERT_NO_MSG(buf != NULL);
+
+	if (IS_ENABLED(CONFIG_LOCATION_METHOD_CELLULAR) && cell_data != NULL) {
+		err = cloud_service_here_rest_pos_req_encode_cellular(cell_data, &cell_buf);
+		if (err) {
+			LOG_ERR("Too small buffer for HTTP request body");
+			goto end;
+		}
+	}
+
+	if (IS_ENABLED(CONFIG_LOCATION_METHOD_WIFI) && wifi_data != NULL) {
+		err = cloud_service_here_rest_pos_req_encode_wifi(wifi_data, &wifi_buf);
+		if (err) {
+			LOG_ERR("Failed to generate Wi-Fi positioning request, err: %d", err);
+			goto end;
+		}
+	}
+
+	__ASSERT_NO_MSG(cell_buf != NULL || wifi_buf != NULL);
+	int cell_buf_len = (cell_buf != NULL) ? strlen(cell_buf) : 0;
+	int wifi_buf_len = (wifi_buf != NULL) ? strlen(wifi_buf) : 0;
+
+	*buf = k_malloc(cell_buf_len + wifi_buf_len + 1);
+	if (*buf == NULL) {
+		LOG_ERR("Out of memory when reserving HTTP request body");
+		err = -ENOMEM;
+		goto end;
+	}
+
+	(*buf)[0] = '\0';
+
+	if (cell_buf != NULL) {
+		strcpy(*buf, cell_buf);
+	}
+	if (wifi_buf != NULL) {
+		if (cell_data != NULL) {
+			/* If there are both cellular and Wi-Fi data,
+			 * remove last '}' from encoded cellular data and
+			 * replace first '{' with ',' in Wi-Fi data.
+			 */
+			wifi_buf[0] = ',';
+			(*buf)[strlen(*buf) - 1] = '\0';
```

Comment:
You could also use `cell_buf_len` here instead of calling `strlen()`.

Reply:
Yep, I used `buf` to avoid future mistakes with this code but I can change it.